### PR TITLE
Add stream SSTable API to stream SSTable components

### PR DESCRIPTION
**Pull request by sarankk**
We want to create an API for streaming SSTable components using Vertx’s zero copy sendFile API

### CODE REVIEW

File: src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.google.common.util.concurrent;
+
+/**
+ * Custom guava Rate Limiter, uses SmoothBursty Ratelimiter
+ */
+public class SidecarRateLimiter
```
Comment (yifan-c) [inline src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java:7]: It simple wraps a RateLimiter and delegate all calls to it. The class seems unnecessary. Can you elaborate?

File: src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    public void setRate(final double permitsPerSecond)
+    {
+        this.rateLimiter.setRate(permitsPerSecond);
+    }
+
+    public double getRate()
+    {
+        return this.rateLimiter.getRate();
+    }
```
Comment (yifan-c) [inline src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java]: Rate is only set via `Configuration`, the setter provided is not used and unnecessary. 
Same for the getter.

File: src/main/java/org/apache/cassandra/sidecar/Configuration.java [hunk 3]
```diff
@@ -255,10 +291,17 @@ public Builder setSslEnabled(boolean enabled)
...
+            return this;
+        }
+
         public Configuration build()
         {
-            return new Configuration(cassandraHost, cassandraPort, host, port, healthCheckFrequencyMillis, isSslEnabled,
-                                     keyStorePath, keyStorePassword, trustStorePath, trustStorePassword);
+            return new Configuration(cassandraHost, cassandraPort, cassandraDataDirs, host, port,
+                    healthCheckFrequencyMillis, isSslEnabled, keyStorePath, keyStorePassword, trustStorePath,
+                    trustStorePassword, rateLimitStreamRequestsPerSecond);
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/Configuration.java]:
```suggestion
            return new Configuration(cassandraHost, cassandraPort, cassandraDataDirs, host, port,
                                     healthCheckFrequencyMillis, isSslEnabled, keyStorePath, keyStorePassword, trustStorePath,
                                     trustStorePassword, rateLimitStreamRequestsPerSecond);
```

File: src/main/java/org/apache/cassandra/sidecar/exceptions/RangeException.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package org.apache.cassandra.sidecar.exceptions;
+
+/**
+ * Custom exception
+ */
+public class RangeException extends RuntimeException
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/exceptions/RangeException.java:6]: Constructors like `public RangeException()` and `public RangeException(Throwable cause)` are not used.

File: src/main/java/org/apache/cassandra/sidecar/models/HttpResponse.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        response.setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code()).setStatusMessage(msg).end();
+    }
+
+    public void sendFile(File file)
+    {
+        response.putHeader(HttpHeaderNames.CONTENT_TYPE, HttpHeaderValues.APPLICATION_OCTET_STREAM)
+                .putHeader(HttpHeaderNames.CONTENT_LENGTH, Long.toString(file.length()))
+                .sendFile(file.getAbsolutePath());
+    }
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/models/HttpResponse.java:71]: Those 2 methods are not used.

File: src/main/java/org/apache/cassandra/sidecar/models/Range.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        return new Range(Math.max(start, range.start()), Math.min(end, range.end()));
+    }
+
+    /**
+     * Accepted string formats "start-end", both ends of the range required to be parsed
+     * Sample accepted formats "1-2", "232-2355"
+     */
+    public static Range parseAbsolute(@NotNull String range)
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/models/Range.java]: How about making `parseAbsolute` a private static function? so that there is no need to check if wrong method was used. 
Providing a single `parse` function makes it easier to use too.

File: src/main/java/org/apache/cassandra/sidecar/models/Range.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+import org.apache.cassandra.sidecar.exceptions.RangeException;
+
+/**
+ * Accepted Range formats are start-end, start-, -suffix_length
+ * start-end (start = start index of the range, end = end index of the range, both inclusive)
+ * start- (start = start index of the range, end = end of file)
+ * -suffix-length (Requested length from end of file)
+ */
+public class Range
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/models/Range.java:18]: Parse the string representation of a range seems clumsy. How about defining header fields dedicated for START, END, and SUFFIX-LENGTH? So there is no need to extract fields using regular expression.

File: src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            }
+        }
+    }
+
+    /**
+     * Enum to hold key length values
+     */
+    public enum KeyLength
+    {
+        ONE, TWO, THREE, FOUR
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java]: nit: it is only used to represent 1, 2, 3, 4... declaring enum types seem overly complicated.

File: src/main/java/org/apache/cassandra/sidecar/utils/FileStreamer.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+ * General handler for serving files
+ */
+@Singleton
+public class FileStreamer
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileStreamer.class);
+    private static final ScheduledExecutorService SCHEDULER = Executors.newScheduledThreadPool(1,
+            new ThreadFactoryBuilder().setNameFormat("acquirePermit").setDaemon(true).build());
+    private static final Duration DELAY = Duration.ofSeconds(5);
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/utils/FileStreamer.java]: Maybe make them configurable.

File: src/test/java/org/apache/cassandra/sidecar/AbstractHealthServiceTest.java [hunk 10]
```diff
@@ -129,9 +130,9 @@ private WebClientOptions getWebClientOptions()
         return options;
     }
 
-    @DisplayName("Should return HTTP 503 Failure when check=False")
+    @DisplayName("Should return HTTP 200 OK when check=False")
```
Comment (yifan-c) [inline src/test/java/org/apache/cassandra/sidecar/AbstractHealthServiceTest.java]: Wrong display name. "200 but NOT_OK"
But.. if the changes to the health endpoint is not necessary for the patch. Please just revert.

File: src/main/java/org/apache/cassandra/sidecar/routes/HealthService.java [hunk 11]
```diff
@@ -53,18 +52,13 @@ public HealthService(CassandraAdapterDelegate cassandra)
     }
 
     @Operation(summary = "Health Check for Cassandra's status",
-    description = "Returns HTTP 200 if Cassandra is available, 503 otherwise",
-    responses = {
-    @ApiResponse(responseCode = "200", description = "Cassandra is available"),
-    @ApiResponse(responseCode = "503", description = "Cassandra is not available")
-    })
+    description = "Returns HTTP 200 if Sidecar is available")
```
Comment (yifan-c) [inline src/main/java/org/apache/cassandra/sidecar/routes/HealthService.java]: The changes in this file do not seem to be related with the patch?

**Review by yifan-c — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/cassandra/sidecar/models/Range.java (hunk 7 above)
Comment (dineshjoshi) [reply src/main/java/org/apache/cassandra/sidecar/models/Range.java:18]: The [range header is a HTTP standard](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range). Its best to stick to standards as standard HTTP clients can be used to interact with this API.

File: src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java (hunk 1 above)
Comment (sarankk) [reply src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java:7]: @yifan-c we wanted to use `this.rateLimiter.queryEarliestAvailable(nowMicros);` method to get time to wait, this method is package protected hence using this wrapper class

File: src/main/java/org/apache/cassandra/sidecar/exceptions/RangeException.java (hunk 4 above)
Comment (sarankk) [reply src/main/java/org/apache/cassandra/sidecar/exceptions/RangeException.java:6]: Removed them

File: src/main/java/org/apache/cassandra/sidecar/models/HttpResponse.java (hunk 5 above)
Comment (sarankk) [reply src/main/java/org/apache/cassandra/sidecar/models/HttpResponse.java:71]: We might have to set `InternalServerError` in other handlers and likewise `sendFile` method might be directly called if range of file to be streamed is not known

File: src/main/java/org/apache/cassandra/sidecar/routes/HealthService.java (hunk 11 above)
Comment (sarankk) [reply src/main/java/org/apache/cassandra/sidecar/routes/HealthService.java]: @yifan-c We wanted to avoid having sidecar's health response dependent on Cassandra's health status. Since the changes needed were minimal included these along with this patch

File: src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java (hunk 8 above)
Comment (sarankk) [reply src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java]: Mainly added to improve readability of this switch statement https://github.com/apache/cassandra-sidecar/blob/0083e0d62dd0a9aa63fa977ffeea6d7609f387c3/src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java#L76

Comment (yifan-c) [reply src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java]: Numbers are not best for readability. Please declare the enum types that are more descriptive. 
For example, `ONE` represents `JUST_KEYSPACE`, `TWO` means `KEYSPACE_TABLE`, `THREE` is `KEYSPACE_TABLE_SNAPSHOT`, so on so forth.

File: src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java (hunk 1 above)
Comment (yifan-c) [reply src/main/java/com/google/common/util/concurrent/SidecarRateLimiter.java:7]: OK. Can you add a java doc to clarify?

**Review by yifan-c — APPROVED** [review]

**Review by dineshjoshi — APPROVED** [review]
